12073 product repo get different sku case

josh-carter opened the pull request:
<!--- Provide a general summary of the Pull Request in the Title above -->
### Description
Take an example Product with SKU `test_sku`

In ProductRepository currently, if we're using get() and pass the SKU `TesT_SKu` we would get an undefined index as described inside #12073.

However It makes sense to be able to retrieve product using different cased SKU as the database is case insensitive and SKU is unique. But the PHP is case sensitive, therefore, to fix the problem we can lowercase the SKU in the PHP code.

<!--- Provide a description of the changes proposed in the pull request -->

### Fixed Issues (if relevant)
<!--- Provide a list of fixed issues in the format magento/magento2#<issue_number>, if relevant  -->
1. [magento/magento2#12073: ProductRepository fails to load product with camelcase SKU](https://github.com/magento/magento2/issues/12073)

### Manual testing scenarios
<!--- Provide a set of unambiguous steps to test the proposed code change -->
1. Create Product with SKU `test_sku`
2. Make a call to `ProductRepository::get('tESt_sKu')`
3. Successfully retrieve created Product 

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

josh-carter committed:
Add Test to ProductRepositoryTest to test get with different cased sku

josh-carter committed:
Use lowercased sku for instances array key

- Then use lowercase sku in get() method which references the instances array via array key this way we make sure no matter which case is used to get the product we get it correctly

- This makes sense as the DB is case insensitive but the PHP is not and currently would throw undefined index if trying to get product with sku: test and passing tESt to get()

orlangur commented:
Hi @josh-carter, putting this PR on hold as hope to get fix done in scope of https://github.com/magento/magento2/pull/12169 (slightly different approach: product is cached exactly with the same SKU how it was requested instead of `strtolower` call as it is done in this PR).

josh-carter commented:
@orlangur Ok cool, my thoughts would be, it saves multiple loading of the same product and keeps functionality correct.

orlangur reviewed (changes requested):
Hi @josh-carter, let's continue in this PR :)

Please rework implementation of `cacheProduct` in https://github.com/magento/magento2/commit/cb3b5d1e0715632c15d8a9697538afca4083a859#diff-ff4893a5628d34910295bae32fbcddfdL241 so that code behavior remain unchanged.

Just change implementation of `cacheProduct` in such a way that we pass both `productId` and `sku` instead of obtaining them from product instance.

Another case is that MySQL does not distinguish not just lower/upper case but also е and ё, o and ö, a and ã and so on. Thus there is no need to complicate implementation with `strtolower` or other approaches for key normalization.

josh-carter commented:
@orlangur Ok that's cool with me. I'll get it updated and get this ready for review.

josh-carter committed:
Revert lowecase logic introduced in previous commit

josh-carter committed:
Add optional Sku parameter in cacheProduct function

- Lets us pass the searched for SKU and use as instance key

josh-carter committed:
Fix Unit Tests - Remove Unused getSku() expectated calls

- This is no longer called everytime we call the get() method because we pass the sku searched for when using the cacheProduct method()

josh-carter committed:
Fix Static tests

- New line at end of ProductRepository
- Remove unused local variable in ProductRepositoryTest

orlangur commented on app/code/Magento/Catalog/Model/ProductRepository.php:
```diff
@@ -302,12 +302,14 @@ protected function getCacheKey($data)
      * @return void
      */
-    private function cacheProduct($cacheKey, \Magento\Catalog\Api\Data\ProductInterface $product)
+    private function cacheProduct($cacheKey, \Magento\Catalog\Api\Data\ProductInterface $product, $sku = null)
```
See: https://github.com/magento/magento2/commit/cb3b5d1e0715632c15d8a9697538afca4083a859#diff-ff4893a5628d34910295bae32fbcddfdL241

Please pass `productId` instead of product instance in both cases. Don't add default parameter and some logic into `cacheProduct`, pass both product ID and SKU.

josh-carter committed:
Alter cacheProduct method to always pass ID and SKU

- Also add expected methods to unit tests

josh-carter committed:
Fix static tests move cacheProduct function to multiline arguments

josh-carter committed:
Fix Unit tests make sure getSku methods will return sku string

orlangur commented on app/code/Magento/Catalog/Test/Unit/Model/ProductRepositoryTest.php:
```diff
@@ -313,23 +313,37 @@ public function testGetProductInEditMode()
         $this->assertEquals($this->productMock, $this->model->get($sku, true));
     }
 
     public function testGetBySkuWithSpace()
```
This test does not seem to test initial case anymore. Please change so that it actually checks trimmed SKU.

orlangur commented on app/code/Magento/Catalog/Test/Unit/Model/ProductRepositoryTest.php:
```diff
@@ -313,23 +313,37 @@ public function testGetProductInEditMode()
+        $this->resourceModelMock->expects($this->once())->method('getIdBySku')->with($differentCasedSku)
+            ->will($this->returnValue('test_id'));
+        $this->productMock->expects($this->once())->method('load')->with('test_id');
+        $this->productMock->expects($this->once())->method('getId')->willReturn('test_id');
```
Same here: it looks like we don't check getting product by `'test_sku'` anymore.

ibecmattb commented:
Facing this same issue. How would I go about patching a Magento 2.2.3 install to fix this?

ishakhsuvarov reviewed (changes requested):
I would suggest covering this with different type of test, for instance `api-functional`.

sidolov commented:
@josh-carter , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for collaboration!

orlangur commented:
Sorry @josh-carter, my bad, I was supposed to help you with test here, I remember we started a talk in Slack. Will do it this week.

ishakhsuvarov commented:
@orlangur I'm reopening then, to keep track of it

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=14019) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=14019) before we can accept your contribution.<br/>

orlangur reviewed (changes requested):
Nowadays all changes must be applied to `2.3-develop` first (see https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#rules). Please prepare a new pull request, this one should be `on hold` until the latter is merged.

orlangur commented:
Closing due to inactivity. Feel free to reach me out anytime later if you wish to continue work on this pull request and it will be reopened.
Thanks for collaboration!
